### DISCUSSION TITLE
FastApi is not running every request on same endpoint in separate Thread

### DISCUSSION THREAD

**Discussion by nilansaha** [fastapi/fastapi#6679] [Questions] [answered]
So, I think I understand all `async def` and `def` stuff. I have this piece of code and I am running it with `uvicorn main:app`
```
import time
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def root():
    print("Hitted Root")
    time.sleep(10)
    return {"message": "Hello World"}


@app.get("/hi")
def root_hi():
    print("Hitted Root Hi")
    time.sleep(10)
``` 

If I visit `/hi` and `/` at the same time. The print statement comes instantly and they each end after 10 seconds approximately which must mean that they are starting at the same time in different threads

However if I open two requests to `/hi` the first one ends and the second one then starts i.e. I see the print statement from the first one and then 10 seconds later the print statement from the second one which must mean they are not running on different threads.

I want to know why is that the case and if this is the default behaviour where requests to different endpoints run in different threads but requests to the same endpoint run one after the other. I also wonder if there is a way to make the requests to the same endpoint run in different threads and `at the same time` without using multiple uvicorn workers.

**Comment by Garito** [comment]
time.sleep should be async (which is not) and you should await for it
You could use ```await asyncio.sleep(10)``` instead

**Comment by nilansaha** [comment]
> await asyncio.sleep(10)

I am using normal `def` functions so I shouldn't have to use await and asyncio. Fastapi doc says for normal def functions it runs everything in a separate threadpool

**Comment by Kludex** [comment]
I can't reproduce your behavior. It works just fine here.

**Comment by raphaelauv** [comment]
> I think I understand all async def and def stuff

apparently no

**Comment by nilansaha** [comment]
> > I think I understand all async def and def stuff
> 
> 
> 
> apparently no

Care to elaborate? Or where I am wrong ?

**Comment by David-Lor** [comment, marked as answer]
@nilansaha How are you performing the requests? I found some time ago that (some) browsers (Firefox in my case) seem to avoid performing simultaneous requests to the same endpoint (or localhost?).

Maybe try this code to perform N concurrent requests using Python, I'd say your example would work as expected after trying it:

```python
import threading
import requests

N_REQUESTS = 5

def req():
	requests.get("http://localhost:8000/hi")

threads = [threading.Thread(target=req, daemon=True) for _ in range(N_REQUESTS)]
[th.start() for th in threads]
[th.join() for th in threads]
```

Output from fastapi server:
```
INFO:     Started server process [33905]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
Hitted Root Hi
Hitted Root Hi
Hitted Root Hi
Hitted Root Hi
Hitted Root Hi
INFO:     127.0.0.1:52508 - "GET /hi HTTP/1.1" 200 OK
INFO:     127.0.0.1:52506 - "GET /hi HTTP/1.1" 200 OK
INFO:     127.0.0.1:52510 - "GET /hi HTTP/1.1" 200 OK
INFO:     127.0.0.1:52504 - "GET /hi HTTP/1.1" 200 OK
INFO:     127.0.0.1:52512 - "GET /hi HTTP/1.1" 200 OK
```

**Comment by nilansaha** [comment]
Indeed that is the case. Chrome was not allowing to run another request on same endpoint until the first one is finished. And thats why the confusion.
